cloud/cluster/cluster.go

```diff
@@ -5,121 +5,133 @@ package cluster
 
 import (
 	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/twitter/scoot/common/stats"
 )
 
-// Cluster represents a group of Nodes and has mechanisms for receiving updates.
-type Cluster struct {
-	state    *state
-	reqCh    chan interface{}
-	updateCh chan ClusterUpdate
-	subs     []chan []NodeUpdate
+var ClusterUpdateLoopFrequency time.Duration = time.Duration(250) * time.Millisecond
+
+// Cluster represents a group of Nodes and has mechanism for
+// setting current Node list and receiving correlating Node updates.
+type Cluster interface {
+	RetrieveCurrentNodeUpdates() []NodeUpdate // used by scheduler's cluster state to get updates
+	SetLatestNodesList(nodes []Node)          // used by fetcher to give cluster the list of current nodes
+	GetNodes() []Node
 }
 
-// Clusters can be updated in two ways:
-// *) a new state of the Cluster, which is a []Node
-// *) updates to specific Nodes, which is a []NodeUpdate
-type ClusterUpdate interface{}
+// cluster implementation of Cluster
+type cluster struct {
+	state *state
+
+	stat stats.StatsReceiver
+
+	// latestFetchedNodes, when not nil, contains the latest node list from fetcher.  When
+	// cluster start processing the latestFetchedNodes, it grabs a copy an sets this to nil
+	latestFetchedNodes   []Node
+	latestFetchedNodesMu sync.RWMutex
+
+	// currentNodeUpdates, the node updates that the scheduler needs to process.  When
+	// scheduler gets the contents of currentNodeUpdates, currentNodeUpdates is reset to nil.
+	currentNodeUpdates   []NodeUpdate
+	currentNodeUpdatesMu sync.RWMutex
+
+	priorNodeUpdateTime  time.Time
+	priorFetchUpdateTime time.Time
+}
 
 // Cluster's ch channel accepts []Node and []NodeUpdate types, which then
 // get passed to its state to either SetAndDiff or UpdateAndFilter
-func NewCluster(state []Node, updateCh chan ClusterUpdate) *Cluster {
-	s := makeState(state)
-	c := &Cluster{
-		state:    s,
-		reqCh:    make(chan interface{}),
-		updateCh: updateCh,
-		subs:     nil,
+func NewCluster(stat stats.StatsReceiver, initialNodes []Node) Cluster {
+	s := makeState([]Node{})
```

Comment:
why don't we want to makeState with initialNodes?

Reply:
starting the loop will acquire the first set of nodes.  The code is simpler without having to do an outer fetch.